sequencer/src/reference_tests.rs

```diff
@@ -0,0 +1,245 @@
+            .parse()
+            .unwrap(),
+    }
+}
+
+const REFERENCE_L1_BLOCK_COMMITMENT: &str = "L1BLOCK~4HpzluLK2Isz3RdPNvNrDAyQcWOF2c9JeLZzVNLmfpQ9";
+
+fn reference_chain_config() -> ChainConfig {
+    ChainConfig {
+        chain_id: 0x8a19.into(),
+        max_block_size: 10240,
+        base_fee: 0.into(),
+        fee_contract: Some(Default::default()),
+        fee_recipient: Default::default(),
+    }
+}
+
+const REFERENCE_CHAIN_CONFIG_COMMITMENT: &str =
+    "CHAIN_CONFIG~L6HmMktJbvnEGgpmRrsiYvQmIBstSj9UtDM7eNFFqYFO";
+
+fn reference_fee_info() -> FeeInfo {
+    FeeInfo::new(
+        FeeAccount::from_str("0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266").unwrap(),
+        0,
+    )
+}
+
+const REFERENCE_FEE_INFO_COMMITMENT: &str = "FEE_INFO~xCCeTjJClBtwtOUrnAmT65LNTQGceuyjSJHUFfX6VRXR";
+
+fn reference_header() -> Header {
+    let builder_key = FeeAccount::generated_from_seed_indexed(Default::default(), 0).1;
+    let fee_info = reference_fee_info();
+    let ns_table = reference_ns_table();
+    let payload = <Payload<_> as TestableBlock>::genesis();
+    let payload_commitment = vid_commitment(&payload.encode(), 1);
+    let builder_commitment = payload.builder_commitment(&ns_table);
+    let builder_signature = FeeAccount::sign_fee(
+        &builder_key,
+        fee_info.amount().as_u64().unwrap(),
+        &ns_table,
+        &payload_commitment,
+    )
+    .unwrap();
+
+    let state = ValidatedState::default();
+
+    Header {
+        height: 42,
+        timestamp: 789,
+        l1_head: 124,
+        l1_finalized: Some(reference_l1_block()),
+        payload_commitment,
+        builder_commitment,
+        ns_table,
+        block_merkle_tree_root: state.block_merkle_tree.commitment(),
+        fee_merkle_tree_root: state.fee_merkle_tree.commitment(),
+        fee_info,
+        chain_config: reference_chain_config().into(),
+        builder_signature: Some(builder_signature),
+    }
+}
+
+const REFERENCE_HEADER_COMMITMENT: &str = "BLOCK~6Ol30XYkdKaNFXw0QAkcif18Lk8V8qkC4M81qTlwL707";
+
+fn reference_transaction() -> Transaction {
+    let payload: [u8; 1024] = std::array::from_fn(|i| (i % (u8::MAX as usize)) as u8);
+    Transaction::new(12648430.into(), payload.to_vec())
+}
+
+const REFERENCE_TRANSACTION_COMMITMENT: &str = "TX~jmYCutMVgguprgpZHywPwkehwXfibQx951gh4LSLmfwp";
+
+fn reference_test<T: Committable + Serialize + DeserializeOwned>(
+    name: &str,
+    reference: T,
+    commitment: &str,
+) {
+    setup_logging();
+    setup_backtrace();
+
+    // Load the expected serialization from the repo.
+    let data_dir = Path::new(env!("CARGO_MANIFEST_DIR")).join("../data");
+    let expected_bytes = std::fs::read(data_dir.join(format!("{name}.json"))).unwrap();
+    let expected: Value = serde_json::from_slice(&expected_bytes).unwrap();
+
+    // Check that the reference object matches the expected serialized form.
+    let actual = serde_json::to_value(&reference).unwrap();
```

Comment:
Is there a reason to use `serde_json` rather than `vbs::Serializer::<hotshot_types::constants::Version01>::serialize` that HotShot uses?

It makes the checked in file human-readable, but if I understand correctly, I think only checking roundtrip json serialization would not catch a lot of the more subtle errors (e.g. reordering but not renaming enum variants)

Reply:
[4a30c28](https://github.com/EspressoSystems/espresso-sequencer/pull/1486/commits/4a30c28e6e506da21c61aee8d4ee95869db7e18d)